module/platform/Webots/src/Webots.cpp

```diff
@@ -0,0 +1,656 @@
+                // Connection successful
+                freeaddrinfo(address);
+                return fd;
+            }
+            // Connection was not successful
+            close(fd);
+        }
+
+        // No connection was successful
+        freeaddrinfo(address);
+        log<NUClear::ERROR>(fmt::format("Cannot connect to server: {}:{}", server_name, port));
+        return -1;
+    }
+
+    Webots::Webots(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+        on<Configuration>("webots.yaml").then([this](const Configuration& config) {
+            // Use configuration here from file webots.yaml
+            time_step = config["time_step"].as<int>();
+
+            // clang-format off
+            auto lvl = config["log_level"].as<std::string>();
+            if      (lvl == "TRACE") { this->log_level = NUClear::TRACE; }
+            else if (lvl == "DEBUG") { this->log_level = NUClear::DEBUG; }
+            else if (lvl == "INFO")  { this->log_level = NUClear::INFO; }
+            else if (lvl == "WARN")  { this->log_level = NUClear::WARN; }
+            else if (lvl == "ERROR") { this->log_level = NUClear::ERROR; }
+            else if (lvl == "FATAL") { this->log_level = NUClear::FATAL; }
+            // clang-format on
+
+            on<Watchdog<Webots, 5, std::chrono::seconds>>().then([this, config] {
+                // We haven't received any messages lately
+                log<NUClear::ERROR>("Connection timed out.");
+                setup_connection(config["server_address"].as<std::string>(), config["port"].as<std::string>());
+            });
+
+            // Prime these this reaction, so when only one ServoTargets is present, at least we have the other
+            emit(std::make_unique<DarwinSensors>());
+
+            // Connect to the server
+            setup_connection(config["server_address"].as<std::string>(), config["port"].as<std::string>());
+        });
+    }
+
+    void Webots::setup_connection(const std::string& server_address, const std::string& port) {
+        // Unbind any previous reaction handles
+        read_io.unbind();
+        send_io.unbind();
+        error_io.unbind();
+        shutdown_handle.unbind();
+
+        if (fd != -1) {
+            // Disconnect the fd gracefully
+            shutdown(fd, SHUT_RDWR);
+            close(fd);
+        }
+
+        fd = tcpip_connect(server_address, port);
+
+        if (fd == -1) {
+            // Connection failed
+            // TODO(Cameron) Try to reconnect after a delay.
+            // setup_connection(server_address, port);
+            log<NUClear::FATAL>("Quitting due to failed connection attempt");
+            powerplant.shutdown();
+            return;
+        }
+
+        // Initaliase the string with ???????
+        std::string initial_message = std::string(7, '?');
+        const int n                 = recv(fd, initial_message.data(), sizeof(initial_message), 0);
```

Comment:
Make `recv` wait until the entire message has been received (I know its only 7 bytes, but forcing the wait might not be a bad thing)
```suggestion
        const int n                 = recv(fd, initial_message.data(), sizeof(initial_message), MSG_WAITALL);
```